ingress: Add cilium ingress controller implementation

sayboras opened the pull request:
### Description

This PR is based on service-mesh-beta branch done by Jarno and Michi before. Compared to this base branch, few changes are done as per below:

- Refactor ingress watcher logic to separate package due to 2 main reasons:
   - watcher pacakge is quite complicated
   - pave the way for future development
- Refactor service, endpoint and cilium envoy configuration into separate struct and file. This will help for unit tests (added in this PR as well)
- Add feature flag with helm options

### Testing

Testing is done with EKS for below cases:

- [x] HTTP kubernetes ingress
- [x] grpc ingress
- [x] TLS termination.


### Conformance testing

Conformance testing was done with https://github.com/kubernetes-sigs/ingress-controller-conformance with my minikube cluster (for faster feedback). While we are still having some failures, these can be addressed in subsequent PR.

- features/default_backend.feature  (all passed) :heavy_check_mark: 
- features/host_rules.feature (70 passed, 3 failed, 1 skipped) :negative_squared_cross_mark: 
- features/path_rules.feature :spinning-gopher: (29 passed, 3 failed, 5 skipped) :negative_squared_cross_mark: 
- features/ingress_class.feature (all passed) :heavy_check_mark: 
- features/load_balancing.feature (all passed) :heavy_check_mark: 


<details> 
<summary>host rule failure</summary>

```

Failed steps:

  Scenario: An Ingress with a host rule should not route traffic when hostname does not match # features/host_rules.feature:64
    When I send a "GET" request to "http://subdomain.bar.com" # features/host_rules.feature:67
      Error: Get "http://10.98.35.160/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

  Scenario: An Ingress with a wildcard host rule should send traffic to the matching backend service # features/host_rules.feature:70
    When I send a "GET" request to "http://bar.foo.com" # features/host_rules.feature:73
      Error: Get "http://10.103.158.100/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

  Scenario: An Ingress with a wildcard host rule should not route traffic matching no dns label # features/host_rules.feature:84
    When I send a "GET" request to "http://foo.com" # features/host_rules.feature:87
      Error: Get "http://10.103.107.12/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)


5 scenarios (2 passed, 3 failed)
37 steps (29 passed, 3 failed, 5 skipped)
1m45.93343714s
```

</details>

<details>
<summary>path rule failure</summary>

```
Failed steps:

  Scenario: An Ingress with prefix path rules should match each labels string prefix # features/path_rules.feature:161
    Then the response status-code must be 404 # features/path_rules.feature:165
      Error: expected status code 404 but 200 was returned

  Scenario: An Ingress with mixed path rules should send traffic to the matching backend service where Exact is preferred # features/path_rules.feature:174
    And the response must be served by the "foo-exact" service # features/path_rules.feature:179
      Error: expected the request to be served by foo-exact but it was served by foo-prefix

  Scenario: An Ingress with a trailing slashes in a prefix path rule should ignore the trailing slash and send traffic to the matching backend service # features/path_rules.feature:181
    Then the response status-code must be 200 # features/path_rules.feature:185
      Error: expected status code 200 but 404 was returned


16 scenarios (13 passed, 3 failed)
74 steps (70 passed, 3 failed, 1 skipped)
10m16.97361674s

```

</details>



```release-note
Add cilium ingress controller implementation
```

michi-covalent committed:
ingress: Add types related to K8s Ingress

This commit adds K8s types related to Ingress. The current plan is to
implement an ingress controller logic in cilium-operator that translates
ingress definitions to corresponding CiliumEnvoyConfig custom resources.

Ref: #17191

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

michi-covalent committed:
ingress: Generate k8s client and register Ingress types

Re-generate k8s client by running `make generate-k8s-api` and register
Ingress and IngressClass types.

Ref: #17191

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

michi-covalent committed:
ingress: Add service, CEC and endpoint creation

This commit is to add first implementation for service, cilium envoy
configuration and endpoint in ingress controller. Envoy resource name
is prepend with ingress namespace and name, so that multiple ingresses
can be configured at the same time.

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
Signed-off-by: Tam Mach <tam.mach@isovalent.com>

michi-covalent committed:
ingress: Add DeletionTimestamp

This commit is to add DeletionTimestamp to slim metav1 struct, so
that ingress controller or any other controller will have more
granular control on life cycle.

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

jrajahalme committed:
ingress: Update permission for operator cluster roles

Cilium operator will watch ingress and create child resources such
as services, endpoint and cilium envoy config. Hence, extra cluster
role permissions {create, update, delete} are required.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

sayboras committed:
ingress: Add ownerReferences for service and endpoints

This commit is to add ownerReferences for service and endpoint created
by Ingress Controller, so that if the ingress resource get removed, all
child namespace-scoped resources will be cleanup automatically by k8s.

Also, correct the typo on resource name while trying to delete Cilium
Envoy Config resource.

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

michi-covalent commented on operator/pkg/ingress/endpoint.go:
```diff
@@ -0,0 +1,88 @@
+		},
+		Subsets: []v1.EndpointSubset{
+			{
+				Addresses: []v1.EndpointAddress{{IP: "192.192.192.192"}}, // dummy
```
very mysterious

sayboras replied:
as discussed offline, this is related to the below point

> accept service without any backend so that we can get rid of the dummy endpoint that gets created by the ingress controller.

aanm replied:
Sorry, I'm not following, what is this about? Can we have a better comment in the code? Or even move this function into the test package?

JAORMX replied:
Having a more thorough comment would help readers. Could that be added?

aanm replied:
Still missing a comment

sayboras replied:
thanks both, I missed this one, should be ok now.

sayboras committed:
tls: Correct TLS resource name in CEC

As mentioned in below docs, the transport socket must be set with name envoy.transport_sockets.tls
and DownstreamTlsContext in the typed_config.

[1]: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener_components.proto#envoy-v3-api-msg-config-listener-v3-filterchain

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

michi-covalent reviewed (approved):
@sayboras thank you for adding https://github.com/cilium/cilium/pull/18867/commits/ddfb4d9e881f7b5163988b2606228b791586ece7 this is very cool :+1: 

@jrajahalme could you review https://github.com/cilium/cilium/pull/18867/commits/3c6000319d1ebe7fa42b777c71475fbac47f1df2?

aanm commented on operator/pkg/ingress/endpoint.go:
```diff
@@ -0,0 +1,79 @@
+	if !cache.WaitForCacheSync(wait.NeverStop, manager.informer.HasSynced) {
+		return manager, fmt.Errorf("unable to sync ingress endpoint")
+	}
+	log.WithField("existing-endpoints", manager.store.ListKeys()).Info("Ingress endpoint synced")
```
I don't think we need to dump all endpoints synchronized.

aanm commented on operator/pkg/ingress/endpoint.go:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	// setup store and informer only for endpoints having label cilium.io/ingress
+	manager.store, manager.informer = informer.NewInformer(
+		cache.NewFilteredListWatchFromClient(k8s.WatcherClient().CoreV1().RESTClient(), "endpoints",
+			v1.NamespaceAll, func(options *metav1.ListOptions) {
+				options.LabelSelector = ciliumIngressLabelKey
+			}),
+		&slim_corev1.Endpoints{},
+		0,
+		cache.ResourceEventHandlerFuncs{},
+		nil,
+	)
```
We also need a watcher for EndpointSlices. See how Cilium does it, we either watch for Endpoints or for EndpointSlices, not both.

sayboras replied:
the watcher here is only for dummy endpoint owned by Ingress object (label cilium.io/ingress: true). It's just a convenient way to upsert (i.e check from cache then perform update or create accordingly), so I don't think we need to watch endpoints or endpointslice in general. 

There is a discussion to find a way to remove this dummy endpoint as well.

sayboras replied:
tracked in https://github.com/cilium/cilium/issues/19262

aanm commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -0,0 +1,368 @@
+	if !cache.WaitForCacheSync(wait.NeverStop, manager.informer.HasSynced) {
+		return manager, fmt.Errorf("unable to sync envoy configs")
+	}
+	log.WithField("existing-envoy-config", manager.store.ListKeys()).Info("CiliumEnvoyConfig synced")
```
no need to list all keys

aanm commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			BackendServices: backendServices,
+			// TODO(michi) what is this
+			// Ingress:         true,
```
Michi, what is this?

sayboras replied:
right now, this is not used, let me remove this comment, will add it back once we actually need to use this attribute in CEC

aanm commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -0,0 +1,368 @@
+			Namespace: ingress.Namespace,
+			Name:      service,
+		})
+	}
```
We can't iterate `services` because it's a map and the order of `backendServices` will be random every time we call `getBackendServices`. Thus, we need to store the keys of `services` into a slice, sort it, and iterate the sorted slice so that we can guarantee that `backendServices` is always returned in-order.

sayboras replied:
good point :100:

aanm commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -0,0 +1,368 @@
+	if len(ingress.Spec.TLS) > 0 {
+		// TODO(tam) extend to list of tls
+		// just take the first one for now
+		domain := ingress.Spec.TLS[0].Hosts[0]
```
To avoid panics, check if `len(ingress.Spec.TLS[0].Hosts) > 0`

aanm commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			OutlierDetection: &envoy_config_cluster_v3.OutlierDetection{
+				SplitExternalLocalOriginErrors: true,
+				ConsecutiveLocalOriginFailure:  &wrapperspb.UInt32Value{Value: 2},
```
Add a comment stating what's the meaning of `2`

aanm commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -0,0 +1,368 @@
+		domains = []string{
+			rule.Host,
+			// match authority header with port (e.g. "example.com:80")
+			rule.Host + ":*",
```
what if `rule.Host` is an IPv6 address? consider replacing this with `net.JoinHostPort(rule.Host, "*")`

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,524 @@
+	if !cache.WaitForCacheSync(wait.NeverStop, ic.ingressInformer.HasSynced) {
+		return
+	}
+	log.WithField("existing-ingresses", ic.ingressStore.ListKeys()).Info("ingresses synced")
```
no need to list all keys

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,524 @@
+	log.WithField("existing-ingresses", ic.ingressStore.ListKeys()).Info("ingresses synced")
+
+	go ic.serviceManager.Run()
+	wait.Until(ic.controlLoop, time.Second, wait.NeverStop)
```
What is this `wait.Until` doing? :thinking:

sayboras replied:
Good point, getting message from event is blocking op, so this wait until is no op

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,524 @@
+	case ingressAddedEvent:
+		log.WithField("ingress", ev.ingress.Name).Info("Handling ingress added event")
+		err = ic.handleIngressAddedEvent(ev)
+		break
```
this is not C, there's no need for `break`

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,524 @@
+	case ingressUpdatedEvent:
+		log.WithField("ingress", ev.newIngress.Name).Info("handling ingress updated event")
+		err = ic.handleIngressUpdatedEvent(ev)
+		break
```
this is not C, there's no need for `break`

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,524 @@
+	case ingressDeletedEvent:
+		log.WithField("ingress", ev.ingress.Name).Info("Handling ingress deleted event")
+		err = ic.handleIngressDeletedEvent(ev)
+		break
```
this is not C, there's no need for `break`

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,524 @@
+	case ingressServiceUpdatedEvent:
+		log.WithField("service", ev.ingressService.Name).Info("Handling ingress service updated event")
+		err = ic.handleIngressServiceUpdatedEvent(ev)
+		break
```
this is not C, there's no need for `break`

aanm commented on operator/pkg/ingress/service.go:
```diff
@@ -0,0 +1,223 @@
+	if newService != nil {
+		return
+	}
+	sm.queue.Add(serviceUpdatedEvent{oldService: oldService, newService: newService})
```
We could perform a DeepEqual between the old and new service and avoid doing any event handling if the services are the same.

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,524 @@
+		return nil
+	}
+
+	_, err = k8s.Client().CoreV1().Services(ingress.Namespace).Create(context.Background(), svc, metav1.CreateOptions{})
```
Out of curiosity, why are we creating Services? :thinking:

sayboras replied:
The goal of this PR is to keep the same design in feature branch without much changes (unless it's blocking).

In my opinion, we will need LB-type service one way or another to route external traffic to cluster, the current design is Ingress owns {Service,Endpoint,CEC} objects. However, there is a discussion to move LB service outside of Ingress controller:

- LB service will be provisioned and managed by user externally.
- Multiple ingress objects can use the  _same_ LB, one benefit is cost saving, the drawback is about handling path conflicts.

This is related to the point mentioned by Michi above

> right now we create one load balancer for each ingress resource. make it so that multiple ingress resources can share a single load balancer.

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,524 @@
+		return nil
+	}
+
+	_, err = k8s.Client().CoreV1().Endpoints(ingress.Namespace).Create(context.Background(), endpoints, metav1.CreateOptions{})
```
The same comment I have left for Endpoints, the same I'm leaving for EndpointSlices. We need to make sure EndpointSlices are also handled.

sayboras replied:
we can use the above thread for discussing this point.

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,524 @@
+			return nil
+		}
+		// Update existing CEC
+		existingEnvoyConfig.Spec = envoyConfig.Spec
```
Never update objects retrieved from the store. Please deep copy them first.

aanm commented on operator/pkg/ingress/service.go:
```diff
@@ -0,0 +1,223 @@
+	case serviceAddedEvent:
+		log.WithField("service", ev.service.Name).Info("Handling service added event")
+		err = sm.handleServiceAddedEvent(ev)
+		break
```
same comments I've left for Ingress cna be applied here.

aanm commented on operator/pkg/ingress/service.go:
```diff
@@ -0,0 +1,223 @@
+	}
+}
+
+func getServiceForIngress(ingress *slim_networkingv1.Ingress) *v1.Service {
```
Consider moving this to the tests package.

sayboras replied:
can you elaborate a little bit here ?

aanm replied:
I thought this function was only used for tests but I can see that's not the case.

aanm commented on install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml:
```diff
@@ -38,18 +38,29 @@ rules:
   - get
   - list
   - watch
+  - create
+  - update
+  - delete
```
Why do we need to perform these verbs? Please add a comment in the commit or in here.

sayboras replied:
the operator will watch ingress objects, and then create child objects {service, endpoint, cec} accordingly, hence CRUD permissions are required for operator role. 

Commit message is updated accordingly, thanks :)

aanm replied:
Can we put these permissions behind the helm flag `enableIngressController` as well as the `get/list/watch` for secrets?

sayboras committed:
ingress: Perform deep equal check in update handler

This commit is to perform deep equal check before adding update event
into the queue.

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

jrajahalme commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -0,0 +1,371 @@
+				return nil, err
+			}
+			tls[host] = &envoy_config_core_v3.TransportSocket{
+				Name: "envoy.transport_sockets.tls",
```
@sayboras I see you found out that this needs to be "envoy.transport_sockets.tls" instead of just "tls". The bare "tls" name has been deprecated for a long time not, but I was not able to pinpoint the change in Envoy that would have removed the support for this deprecated name.

sayboras replied:
I noticed this during testing only, the current envoy proxy is no longer support 'tls' name, as sooner or later we need to update this name, so I _didn't_ look deeper to pinpoint which commit/version removed it.

kaworu reviewed (approved):
Helm changes LGTM

qmonnet reviewed (approved):
Doc changes look good. I'm not super familiar with the rest of the code so I just had a quick look, but found no red flag.

nbusseneau reviewed (approved):
Only looked at files owned by me, changes LGTM 👍🏻

gandro reviewed (approved):
My owned code looks good. Only did a cursory glace over the K8s watcher code changes.

sayboras committed:
ingress: Add support for Ingress PathType

Ingress PathType possible values are Exact, ImplementationSpecific and
Prefix, we can map to envoy match route as per below:

- Exact -> Exact
- ImplementationSpecific or Prefix -> Prefix

As per k8s doc, if PathType is not specified, the behavior should be same
as ImplementationSpecific, which will be Prefix rule in Envoy.

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

sayboras committed:
ingress: Add support for default backend

This commit is to add support for default backend in Ingress. Basically,
only need to configure backendServices, routes, and clusters in CEC.

Resource lookup is not supported right now.

Testing was done manually and automatically with conformance test

Manual testing
```
$ kg ingress basic-ingress -o json | jq .spec.defaultBackend
{
  "service": {
    "name": "ratings",
    "port": {
      "number": 9080
    }
  }
}

$ curl http://$lb:80/ratings/1
{"id":1,"ratings":{"Reviewer1":5,"Reviewer2":4}}
```

Conformance testing:
```
...
6 scenarios (6 passed)
72 steps (72 passed)
1m0.187075485s

```

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

aanm commented on install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml:
```diff
@@ -72,6 +83,12 @@ rules:
+  - ciliumegressnatpolicies/finalizers
+  - ciliumenvoyconfigs
+  - ciliumenvoyconfigs/status
+  - ciliumenvoyconfigs/finalizers
```
Please add the appropriate verbs here instead of using `'*'`

sayboras replied:
all these resources are cilium related CR, I think we can grant all permissions, is your concern mainly about explicity of verbs?

aanm replied:
@sayboras correct, we need to be explicit about which verbs we actually need.

aanm replied:
@sayboras why are the egressnatpolicies part of this PR? Also, we don't need the /status nor the /finalizers subresource

sayboras replied:
thanks, forced push to address the above comment, below is difft output

```
install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml --- Text
87 87   - ciliumlocalredirectpolicies
88 88   - ciliumlocalredirectpolicies/status
89 89   - ciliumlocalredirectpolicies/finalizers
90 ..   - ciliumegressnatpolicies
91 ..   - ciliumegressnatpolicies/status
92 ..   - ciliumegressnatpolicies/finalizers
93 90   - ciliumenvoyconfigs
94 ..   - ciliumenvoyconfigs/status
95 ..   - ciliumenvoyconfigs/finalizers
96 91   verbs:
97 92   - get
98 93   - list
```

aanm commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -0,0 +1,411 @@
+	a, err := anypb.New(message)
+	if err != nil {
+		panic(err.Error())
+	}
```
Let's not panic. We can write a log.Error, but not panic.

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+		ic.queue.AddRateLimited(event)
+	} else {
+		ic.queue.Forget(event)
+	}
```
Should we forget it? Perhaps we should log an error as well?

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+		WithField("svc-status", service.Status.LoadBalancer).
+		WithField("ingress-status", ingress.Status.LoadBalancer)
+
+	if reflect.DeepEqual(service.Status.LoadBalancer, ingress.Status.LoadBalancer) {
```
`if service.Status.LoadBalancer.DeepEqual(&ingress.Status.LoadBalancer) {`

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+		WithField("ingress-status", ingress.Status.LoadBalancer)
+
+	if reflect.DeepEqual(service.Status.LoadBalancer, ingress.Status.LoadBalancer) {
+		scopedLog.Info("No change in load balancer status")
```
No need to log this in `Info` message, otherwise they will be too verbose

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+		scopedLog.WithError(err).Warn("Failed to update ingress status")
+		return err
+	}
+	scopedLog.Info("Updated ingress status")
```
Same for the Info message, perhaps this should be Debug messages instead.

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+	}
+
+	newIngressStatus := getIngressForStatusUpdate(ingress, service.Status.LoadBalancer)
+	_, err = k8s.Client().NetworkingV1().Ingresses(ingress.Namespace).UpdateStatus(context.Background(), newIngressStatus, metav1.UpdateOptions{})
```
Not blocking but would be great to know if we could use `ApplyStatus` instead perhaps it would even prevent us from having to generate an ingress object with `getIngressForStatusUpdate`? It looks like `ApplyStatus` does not need a full Ingress object, only its name and status.

christarazi replied:
@sayboras Could you respond to Andre question?

sayboras replied:
oh I missed this, let me track this work along with standard informer pattern as per snippet shared by Andre before.

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+		log.WithField("ingress", ev.ingress.Name).Info("Handling ingress deleted event")
+		err = ic.handleIngressDeletedEvent(ev)
+	case ingressServiceUpdatedEvent:
+		log.WithField("service", ev.ingressService.Name).Info("Handling ingress service updated event")
```
Let's change these log fields to `Debug` instead, otherwise the logs will be too verbose.

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+	if exists {
+		// Endpoints already exists in the cache. For now assume that it was created by the ingress
+		// controller.
+		log.WithField("endpoints", key).Info("Endpoints already exists. Continuing...")
```
same here for the `Info` -> `Debug`

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+		if err != nil {
+			log.WithError(err).WithField("ingress", ingress.Name).Error("Failed to update CiliumEnvoyConfig for ingress")
+		}
+		log.WithField("ingress", ingress.Name).Error("Updated CiliumEnvoyConfig for ingress")
```
I don't think this message should be an Error

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+		_, err = k8s.CiliumClient().CiliumV2alpha1().CiliumEnvoyConfigs(ingress.Namespace).Update(context.Background(), newEnvoyConfig, metav1.UpdateOptions{})
+		if err != nil {
+			log.WithError(err).WithField("ingress", ingress.Name).Error("Failed to update CiliumEnvoyConfig for ingress")
+		}
```
should we `return err` for this case?

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+	newIngress := k8s.ObjToV1Ingress(newObj)
+	if newIngress == nil {
+		return
+	}
+	ic.queue.Add(ingressUpdatedEvent{oldIngress: oldIngress, newIngress: newIngress})
```
even before adding something into the queue we could do a DeepEqual and if the objects are equal then we can ignore this event.

aanm commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,447 @@
+		return err
+	}
+	if !exists {
+		log.WithField("cilium-envoy-config", resourceName).Info("CiliumEnvoyConfig already deleted. Continuing...")
```
same for the Info messages. I didn't catch all of them but it seems that a lot of the Info messages could be changed to Debug instead.

aanm commented on operator/pkg/ingress/service.go:
```diff
@@ -0,0 +1,240 @@
+		sm.queue.AddRateLimited(event)
+	} else {
+		sm.queue.Forget(event)
+	}
```
Same here that we should probably log an error for if the event is forgotten

aanm commented on operator/pkg/ingress/service.go:
```diff
@@ -0,0 +1,240 @@
+					Kind:               "Ingress",
+					Name:               ingress.Name,
+					UID:                ingress.UID,
+					BlockOwnerDeletion: boolP(true),
```
we can drop this, see https://github.com/cilium/cilium/pull/19074/commits/37bcb4ddce8dac8776c496bfb7c1dab44cde0a57

aanm commented on pkg/k8s/factory_functions.go:
```diff
@@ -40,6 +40,26 @@ func ObjToV1NetworkPolicy(obj interface{}) *slim_networkingv1.NetworkPolicy {
+		}
+	}
+	log.WithField(logfields.Object, logfields.Repr(obj)).
+		Warn("Ignoring invalid k8s v1 Service")
```
`v1 Service` -> `v1 Ingress`

sayboras committed:
envoy: Avoid having duplicated backend service

The same backend service can be used multiple times in different rules
(e.g. with different host values), which leads to failure in envoy config
due to multiple Cluster resources with same name. Hence, de-dup should be
done.

This is identified as part of conformance test.

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

aanm reviewed (approved):
The only missing piece is this comment https://github.com/cilium/cilium/pull/18867#discussion_r840984965

If this is addressed it's all good from my side.

sayboras committed:
helm: Replace wildcard permission for cilium CRs

This commit is to replace wildcard (e.g. *) permission to explicit verbs
for operator role.

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

christarazi commented on operator/pkg/ingress/envoy_config.go:
```diff
@@ -0,0 +1,373 @@
+		},
+	}
+	if len(ingress.Spec.TLS) > 0 {
+		// TODO(tam) extend to list of tls
+		// just take the first one for now
+		if len(ingress.Spec.TLS[0].Hosts) > 0 {
```
Should we create an issue to track these followups? We have a (loose) policy of not committing new TODOs to the code, and instead tracking them in a dedicated issue and referencing the issue number in the code comments.

sayboras replied:
agreed :100: 

https://github.com/cilium/cilium/issues/19312

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2019-2020 Authors of Cilium
+
+package ingress
```
It would be good to have a godoc explaining what this ingress package is at a high level

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+	ingressService *slim_corev1.Service
+}
+
+// IngressController struct
```
Could we be more descriptive on what IngressController is and what it does?

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+	} else if ic.queue.NumRequeues(event) < ic.maxRetries {
+		ic.queue.AddRateLimited(event)
+	} else {
+		log.Errorf("failed to process event: %s", event)
```
Since this is a log msg that will be read by us or cluster operators, we should add more text for easier debuggability?

```
Failed to process Ingress event ..., skipping
```

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+		return nil
+	}
+	if err := ic.createEnvoyConfig(event.newIngress); err != nil {
+		log.WithError(err).Warn("failed to update CiliumEnvoyConfig")
```
Nit: I notice a difference in formatting the log msgs: some start with capitalization, some don't. Let's be consistent.

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+func (ic *IngressController) handleIngressDeletedEvent(event ingressDeletedEvent) error {
+	// as ciliumEnvoyConfig is non-namespace scoped, it should be removed manually.
+	// other resources (e.g service, endpoints) will be removed automatically via ownerReferences
+	log.WithField("ingress", event.ingress.Name).Debug("Deleting CiliumEnvoyConfig for ingress")
```
Nit: let's create new log fields inside `pkg/logging/logfields` for "ingress" and other commonly used fields in this code

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+	scopedLog := log.WithField("ingress", ingressKey).WithField("service", serviceKey).
+		WithField("svc-status", service.Status.LoadBalancer).
+		WithField("ingress-status", ingress.Status.LoadBalancer)
+
```
Nit: WithFields(logrus.Fields{...})

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+		log.WithField("service", ev.ingressService.Name).Debug("Handling ingress service updated event")
+		err = ic.handleIngressServiceUpdatedEvent(ev)
+	default:
+		err = fmt.Errorf("received an unknown event: %s", ev)
```
%T

christarazi commented on operator/pkg/ingress/service.go:
```diff
@@ -0,0 +1,221 @@
+
+	service, ok := objFromCache.(*slim_corev1.Service)
+	if !ok {
+		return nil, exists, fmt.Errorf("got invalid object from cache")
```
```
unexpected type found in service cache: %T
```

christarazi commented on operator/pkg/ingress/endpoint.go:
```diff
@@ -0,0 +1,81 @@
+
+	endpoint, ok := objFromCache.(*slim_corev1.Endpoints)
+	if !ok {
+		return nil, exists, fmt.Errorf("got invalid object from cache")
```
Same here and the rest of the `getByKey` functions

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+	ingressKey := getIngressKeyForService(service)
+	ingress, err := ic.getByKey(ingressKey)
+	if err != nil {
+		log.WithError(err).WithField("service", serviceKey).Warn("Failed to lookup ingress for service")
```
I don't see the need to add a warning log msg here. We'll return an error from here anyway and the caller will log with enough context for why we failed to handle it. This comment applies to all other `handle*Event` for the `getByKey` calls

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+	svc := getServiceForIngress(ingress)
+	svcKey, err := cache.MetaNamespaceKeyFunc(svc)
+	if err != nil {
+		log.Warn("failed to get service key for ingress")
```
This seems unnecessary too

christarazi commented on operator/pkg/ingress/ingress.go:
```diff
@@ -0,0 +1,446 @@
+	endpoints := getEndpointsForIngress(ingress)
+	key, err := cache.MetaNamespaceKeyFunc(endpoints)
+	if err != nil {
+		log.WithError(err).Warn("MetaNamespaceKeyFunc returned an error")
```
Same here (unnecessary log)

christarazi commented on operator/pkg/ingress/service.go:
```diff
@@ -0,0 +1,221 @@
+	ingressQueue workqueue.RateLimitingInterface
+}
+
+func newServiceManager(ingressQueue workqueue.RateLimitingInterface, maxRetries int) (*serviceManager, error) {
+	manager := &serviceManager{
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		ingressQueue: ingressQueue,
+		maxRetries:   maxRetries,
+	}
+
+	manager.store, manager.informer = informer.NewInformer(
+		cache.NewFilteredListWatchFromClient(k8s.WatcherClient().CoreV1().RESTClient(), "services",
+			v1.NamespaceAll, func(options *metav1.ListOptions) {
+				options.LabelSelector = ciliumIngressLabelKey
+			}),
+		&slim_corev1.Service{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    manager.handleAddService,
+			UpdateFunc: manager.handleUpdateService,
+			DeleteFunc: manager.handleDeleteService,
+		},
+		nil,
+	)
```
We already have a services informer inside `operator/watchers/k8s_service_sync.go`, so this is duplicating all the services in the cluster inside the Operator memory-wise. Instead, we can have one informer with multiple subscribers. Currently, we are limited by not being able to create SharedInformers (which share a memory cache) because they don't support our slim structures yet (cc @aanm). 

In the meantime, it should be possible to convert the logic here to the subscriber model a la `serviceCacheSubscriber` inside `operator/watchers/service_cache.go`. In the new subscriber object, we can store the queue if needed and pass a reference to the existing service store as well.

Ref: https://github.com/cilium/cilium/issues/15471

aanm replied:
@christarazi let's leave this as is because `operator/watchers/service_cache.go` is conditional if kvstore is enabled, `synchronize-k8s-services` is set and this informer has a label selector.

christarazi replied:
Ah thanks, I didn't notice the label selector

christarazi commented on operator/pkg/ingress/service.go:
```diff
@@ -0,0 +1,221 @@
+			Name:     "http",
+			Protocol: "TCP",
+			Port:     80,
+			// TODO(michi) how do we deal with multiple target ports?
+			// TODO(jarno) It seems that when service is accessed via
+			//             NodePort all these targetports are ignored anyway,
+			//             and the targetport defined in the backend service is
+			//             used regardless what is specified here.
```
Track in follow up issue?

sayboras replied:
agreed https://github.com/cilium/cilium/issues/19312

christarazi commented on pkg/k8s/apis/cilium.io/v2alpha1/cec_types.go:
```diff
@@ -11,7 +11,7 @@ import (
 	"google.golang.org/protobuf/encoding/prototext"
 	"google.golang.org/protobuf/proto"
-	anypb "google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/anypb"
```
I don't understand this commit :thinking:?

sayboras replied:
oh i overlooked this one, originally it was for namespace change, however, this was embedded in CEC PR as per below comment. Let me drop this commit.

https://github.com/cilium/cilium/pull/18894#discussion_r839472805

christarazi commented on operator/pkg/ingress/ingress_options.go:
```diff
@@ -5,30 +5,18 @@ package ingress
-// which disables Hubble export.
+// WithMaxRetries sets the maximum number of retries while processing events
 func WithMaxRetries(maxRetries int) Option {
 	return func(o *Options) error {
```
I think these changes shouldn't be a part of this commit (`ingress: Add feature flag for ingress controller `)?

sayboras replied:
you are right, let me split it out

christarazi commented on operator/pkg/ingress/service.go:
```diff
@@ -141,13 +144,11 @@ func (sm *serviceManager) handleEvent(event interface{}) error {
-		break
 	case serviceDeletedEvent:
-		break
+		//doing nothing right now
```
Unnecessary code comment IMO. Should this be a TODO that's tracked in an issue?

sayboras replied:
as we are having ownerReferences, the delete event case here is just to avoid default case for unknown event.

christarazi replied:
I think that's fine, but I was trying to get it making the comment a bit more useful to explain why we aren't doing anything :)

christarazi reviewed (commented):
I reviewed the code at a high-level. There were some areas related to Envoy which I don't understand so I essentially skimmed through this. I have identified a few code style nits for consistency and improvements to logging to prevent unnecessary noise and unhelpful msgs. I didn't review too closely the core K8s event handling logic (as Andre handled that), but instead focused on integrating with our existing K8s watcher infrastructure. The biggest concern I have with the PR is that we will be duplicating K8s services in the Operator memory-wise. I believe everything else is non-blocking.

sayboras committed:
ingress: Add feature flag for ingress controller

This commit is to add feature flags for enabling cilium ingress controller.
From agent side, enable-envoy-config is used to enable processing of
CiliumEnvoyConfig CR. Helm option and config map are updated with default
value as false.

Signed-off-by: Tam Mach <tam.mach@isovalent.com>
